pkg/network/ebpf/c/prebuilt/tracer.c

```diff
@@ -820,6 +850,74 @@ int kprobe__ip_make_skb(struct pt_regs* ctx) {
     return 0;
 }
 
+#if ROUTE_LOOKUP_USE_EBPF
+
+SEC("kprobe/ip_route_output_flow")
+int kprobe__ip_route_output_flow(struct pt_regs* ctx) {
+    struct net *net = (struct net*) PT_REGS_PARM1(ctx);
+    struct flowi4* fl4 = (struct flowi4*) PT_REGS_PARM2(ctx);
+    struct sock* sk = (struct sock*) PT_REGS_PARM3(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+
+    ip_route_flow_t flow = {};
+    flow.sk = sk;
+    flow.fl = fl4;
+    bpf_probe_read(&flow.net_ns, sizeof(flow.net_ns), ((char*)net) + offset_ino());
+    bpf_map_update_elem(&ip_route_output_flows, &pid_tgid, &flow, BPF_ANY);
+    log_debug("kprobe/ip_route_output_flow: pid_tgid: %d\n", pid_tgid);
+
+    return 0;
+}
+
+SEC("kretprobe/ip_route_output_flow")
+int kretprobe__ip_route_output_flow(struct pt_regs* ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+
+    // Retrieve socket pointer from kprobe via pid/tgid
+    ip_route_flow_t *flow = (ip_route_flow_t*) bpf_map_lookup_elem(&ip_route_output_flows, &pid_tgid);
+    if (!flow) {
+        return 0;
+    }
+
+    // Make sure we clean up that pointer reference
+    bpf_map_delete_elem(&ip_route_output_flows, &pid_tgid);
+
+    struct rtable *rt = (struct rtable*)PT_REGS_RC(ctx);
+    if (IS_ERR_OR_NULL(rt)) {
+        log_debug("kretprobe/ip_route_output_flow: route is not available pid_tgid=%d\n", pid_tgid);
+        return 0;
+    }
+
+    struct flowi4 *fl4 = (struct flowi4*) flow->fl;
+
+    dest_tuple_t dest = {};
+    dest.saddr_h = 0;
+    dest.daddr_h = 0;
+    bpf_probe_read(&dest.saddr_l, sizeof(__u32), ((char*)fl4) + offset_saddr_fl4());
+    bpf_probe_read(&dest.daddr_l, sizeof(__u32), ((char*)fl4) + offset_daddr_fl4());
```

Comment:
Since this kprobe is  already relying on non-guessed offsets, we shouldn't mix direct field access and the more confusing offset-guessing based accesses in this kprobe.

When this kprobe is loaded, we'll be able to write:

```
bpf_probe_read(&dest.daddr_l, sizeof(__u32), fl4->daddr);
```

no?